File: app/src/main/java/com/cwlarson/deviceid/database/Hardware.java
package com.cwlarson.deviceid.database;

import android.app.Activity;
import android.app.ActivityManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.AsyncTask;
import android.os.BatteryManager;
import android.os.Build;
import android.support.v4.content.ContextCompat;
import android.text.format.Formatter;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.Display;
import android.view.WindowManager;

import com.cwlarson.deviceid.R;
import com.cwlarson.deviceid.databinding.ChartItem;
import com.cwlarson.deviceid.databinding.Item;
import com.cwlarson.deviceid.databinding.ItemType;

import java.io.File;
import java.lang.reflect.Method;

import static android.content.Context.ACTIVITY_SERVICE;

class Hardware {
    private final String TAG = Hardware.class.getSimpleName();
    private final Context context;

    Hardware(Activity activity, AppDatabase db){
        this.context = activity.getApplicationContext();
        //Set Hardware Tiles
        ItemAdder itemAdder = new ItemAdder(context, db);
        itemAdder.addItems(getDeviceScreenDensity());
        itemAdder.addItems(getRamSize());
        itemAdder.addItems(getFormattedInternalMemory());
        itemAdder.addItems(getFormattedExternalMemory());
        getBattery();
    }

    private Item getDeviceScreenDensity() {
        Item item = new Item("Screen Density", ItemType.HARDWARE);
        int density = context.getResources().getDisplayMetrics().densityDpi;
        int width = 0, height = 0;
        final DisplayMetrics metrics = new DisplayMetrics();
        WindowManager windowManager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = windowManager.getDefaultDisplay();
        Method mGetRawH, mGetRawW;

        try {
            // For JellyBean 4.2 (API 17) and onward
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN_MR1) {
                display.getRealMetrics(metrics);

                width = metrics.widthPixels;
                height = metrics.heightPixels;
            } else {
                mGetRawH = Display.class.getMethod("getRawHeight");
                mGetRawW = Display.class.getMethod("getRawWidth");

                try {
                    width = (Integer) mGetRawW.invoke(display);
                    height = (Integer) mGetRawH.invoke(display);
                } catch (Exception e) {
                    Log.e(TAG, "IllegalArgumentException in getDeviceScreenDensity");
                }
            }
        } catch (Exception e3) {
            Log.e(TAG, "NoSuchMethodException in getDeviceScreenDensity");
        }
        String sizeInPixels = " ("+Integer.toString(height)+"x"+Integer.toString(width)+")";

        switch (density) {
            case DisplayMetrics.DENSITY_LOW:
                item.setSubtitle("LDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_MEDIUM:
                item.setSubtitle("MDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_HIGH:
                item.setSubtitle("HDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_XHIGH:
                item.setSubtitle("XHDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_XXHIGH:
                item.setSubtitle("XXHDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_XXXHIGH:
                item.setSubtitle("XXXHDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_TV:
                item.setSubtitle("TVDPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_400:
                item.setSubtitle("400DPI"+sizeInPixels);
                break;
            case DisplayMetrics.DENSITY_560:
                item.setSubtitle("560DPI"+sizeInPixels);
                break;
            default:
                break;
        }
        return item;
    }

    private Item getRamSize() {
        Item item = new Item("Memory", ItemType.HARDWARE);
        try {
            ActivityManager.MemoryInfo mi = new ActivityManager.MemoryInfo();
            ActivityManager activityManager = (ActivityManager) context.getSystemService(ACTIVITY_SERVICE);
            activityManager.getMemoryInfo(mi);
            if(Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
                //Total memory is only available on API 16+
                item.setSubtitle(context.getResources().getString(R.string
                    .hardware_storage_output_format_api15,
                    Formatter.formatFileSize(context, mi.availMem)));
                item.setChartitem(new ChartItem(mi.availMem,0, R.drawable.ic_memory));
            }else {
                item.setSubtitle(mi.totalMem <= 0 ? context.getResources().getString(R.string.not_found) : context.getResources
                    ().getString(R.string.hardware_storage_output_format,
                    Formatter.formatFileSize(context, mi.availMem),
                    Formatter.formatFileSize(context, mi.totalMem)));
                item.setChartitem(new ChartItem(mi.availMem, mi.totalMem, R.drawable.ic_memory));
            }
        } catch (Exception e) {
            Log.w(TAG, "Exception in getRamSize");
        }
        return item;
    }

    private Item getFormattedInternalMemory() {
        Item item = new Item("Internal Storage", ItemType.HARDWARE);
        long available = new File(context.getFilesDir().getAbsoluteFile().toString()).getFreeSpace();
        long total = new File(context.getFilesDir().getAbsoluteFile().toString()).getTotalSpace();
        if(total>0L)
            item.setSubtitle(context.getResources().getString(R.string.hardware_storage_output_format, Formatter.formatFileSize(context,available), Formatter.formatFileSize(context,total)));
        item.setChartitem(new ChartItem(available,total,R.drawable.ic_storage));
        return item;
    }

    private Item getFormattedExternalMemory() {
        Item item =  new Item("External Storage", ItemType.HARDWARE);
        long availSize=0L,totalSize=0L;
        File[] appsDir = ContextCompat.getExternalFilesDirs(context,null);
        for(File file : appsDir) {
            availSize += file.getParentFile().getParentFile().getParentFile().getParentFile().getFreeSpace();
            totalSize += file.getParentFile().getParentFile().getParentFile().getParentFile().getTotalSpace();
        }
        availSize-=new File(context.getFilesDir().getAbsoluteFile().toString()).getFreeSpace();
        totalSize-=new File(context.getFilesDir().getAbsoluteFile().toString()).getTotalSpace();
        if(totalSize>0L)
            item.setSubtitle(context.getResources().getString(R.string.hardware_storage_output_format, Formatter.formatFileSize(context, availSize), Formatter.formatFileSize(context, totalSize)));
        item.setChartitem(new ChartItem(availSize,totalSize,R.drawable.ic_storage));
        return item;
    }

    private void getBattery() {
        //Item item = new Item("Battery", ItemType.HARDWARE);
        context.registerReceiver(new InfoReceiver(),new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
    }
    //int temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)/10;
    // Are we charging / charged?
    // How are we charging?
    //subTitle = String.valueOf(temperature+"\u00b0C");
    private class InfoReceiver extends BroadcastReceiver {
        @Override
        public void onReceive(Context context, Intent intent) {
            if(intent==null || intent.getAction()==null) return;
            final PendingResult result = goAsync();
            new ProcessInfoBroadcastAsync(context,intent,result).execute();
            context.unregisterReceiver(this);
        }
    }
    private static class ProcessInfoBroadcastAsync extends AsyncTask<Void, Void, Void> {
        private final Context context;
        private final Intent intent;
        private final BroadcastReceiver.PendingResult result;

        ProcessInfoBroadcastAsync(Context context, Intent intent, BroadcastReceiver
            .PendingResult result) {
            this.context = context;
            this.intent = intent;
            this.result = result;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            if(intent.getAction().equalsIgnoreCase(Intent.ACTION_BATTERY_CHANGED)) {
                //int temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)/10;
                int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
                StringBuilder subtitle = new StringBuilder();
                subtitle.append(level).append("% - ");
                // Are we charging / charged?
                switch (intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
                    case BatteryManager.BATTERY_STATUS_CHARGING:
                        subtitle.append(context.getString(R.string.BATTERY_STATUS_CHARGING));
                        break;
                    case BatteryManager.BATTERY_STATUS_FULL:
                        subtitle.append(context.getString(R.string.BATTERY_STATUS_FULL));
                        break;
                    case BatteryManager.BATTERY_STATUS_DISCHARGING:
                        subtitle.append(context.getString(R.string.BATTERY_STATUS_DISCHARGING));
                        break;
                    case BatteryManager.BATTERY_STATUS_NOT_CHARGING:
                        subtitle.append(context.getString(R.string.BATTERY_STATUS_NOT_CHARGING));
                        break;
                    default:
                        subtitle.append(context.getString(R.string.BATTERY_STATUS_UNKNOWN));
                        break;
                }
                // How are we charging?
                int chargePlug = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1);
                boolean usbCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_USB;
                boolean acCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_AC;
                subtitle.append(" ");
                if(usbCharge)
                    subtitle.append(context.getString(R.string.BATTERY_PLUGGED_USB));
                else if(acCharge)
                    subtitle.append(context.getString(R.string.BATTERY_PLUGGED_AC));
                //subTitle = String.valueOf(temperature+"\u00b0C");
                Item item = new Item("Battery", ItemType.HARDWARE);
                item.setSubtitle(subtitle.toString());
                item.setChartitem(new ChartItem(100-level,100,R.drawable.ic_battery));
                AppDatabase.getDatabase(context).itemDao().insertItems(item);
                // Must call finish() so the BroadcastReceiver can be recycled.
                result.finish();
            }
            return null;
        }
    }
}

File: app/src/main/java/com/cwlarson/deviceid/database/Hardware.kt
package com.cwlarson.deviceid.database

import android.app.ActivityManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import android.os.Build
import android.os.Environment
import android.os.StatFs
import android.text.format.Formatter
import android.util.DisplayMetrics
import android.view.Display
import androidx.core.content.ContextCompat
import androidx.core.os.EnvironmentCompat
import com.cwlarson.deviceid.R
import com.cwlarson.deviceid.databinding.ChartItem
import com.cwlarson.deviceid.databinding.Item
import com.cwlarson.deviceid.databinding.ItemType
import com.cwlarson.deviceid.util.activityManager
import com.cwlarson.deviceid.util.windowManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import timber.log.Timber
import java.io.File
import java.lang.reflect.Method
import kotlin.coroutines.resume

class Hardware(private val context: Context, private val db: AppDatabase, private val scope: CoroutineScope) {
    init {
        scope.launch {
            db.addItems(context, deviceScreenDensity(), ramSize(), formattedInternalMemory(),
                    formattedExternalMemory(), getBattery())
        }
    }

    private fun deviceScreenDensity() = Item("Screen Density", ItemType.HARDWARE).apply {
        try {
            val density = context.resources.displayMetrics.densityDpi
            val width: Int
            val height: Int
            val metrics = DisplayMetrics()
            val display = context.windowManager.defaultDisplay
            val mGetRawH: Method
            val mGetRawW: Method
            // For JellyBean 4.2 (API 17) and onward
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                display.getRealMetrics(metrics)
                width = metrics.widthPixels
                height = metrics.heightPixels
            } else {
                mGetRawH = Display::class.java.getMethod("getRawHeight")
                mGetRawW = Display::class.java.getMethod("getRawWidth")
                width = mGetRawW.invoke(display) as Int
                height = mGetRawH.invoke(display) as Int
            }
            subtitle = when (density) {
                DisplayMetrics.DENSITY_LOW -> "LDPI"
                DisplayMetrics.DENSITY_MEDIUM -> "MDPI"
                DisplayMetrics.DENSITY_HIGH -> "HDPI"
                DisplayMetrics.DENSITY_XHIGH -> "XHDPI"
                DisplayMetrics.DENSITY_XXHIGH -> "XXHDPI"
                DisplayMetrics.DENSITY_XXXHIGH -> "XXXHDPI"
                DisplayMetrics.DENSITY_TV -> "TVDPI"
                DisplayMetrics.DENSITY_140 -> "140 DPI"
                DisplayMetrics.DENSITY_180 -> "180 DPI"
                DisplayMetrics.DENSITY_200 -> "200 DPI"
                DisplayMetrics.DENSITY_220 -> "220 DPI"
                DisplayMetrics.DENSITY_260 -> "260 DPI"
                DisplayMetrics.DENSITY_280 -> "280 DPI"
                DisplayMetrics.DENSITY_300 -> "300 DPI"
                DisplayMetrics.DENSITY_340 -> "340 DPI"
                DisplayMetrics.DENSITY_360 -> "360 DPI"
                DisplayMetrics.DENSITY_400 -> "400 DPI"
                DisplayMetrics.DENSITY_420 -> "420 DPI"
                DisplayMetrics.DENSITY_440 -> "440 DPI"
                DisplayMetrics.DENSITY_560 -> "560 DPI"
                DisplayMetrics.DENSITY_600 -> "600 DPI"
                else -> { null }
            }.plus(" (${height}x$width)").trim()
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    //Total memory is only available on API 16+
    private fun ramSize() = Item("Memory", ItemType.HARDWARE).apply {
        try {
            val mi = ActivityManager.MemoryInfo()
            context.activityManager.getMemoryInfo(mi)
            if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
                subtitle = context.resources.getString(R.string
                        .hardware_storage_output_format_api15,
                        Formatter.formatFileSize(context, mi.availMem))
                chartItem = ChartItem(mi.availMem.toFloat(), 0f, R.drawable.ic_memory)
            } else {
                subtitle = if (mi.totalMem <= 0)
                    context.resources.getString(R.string.not_found)
                else
                    context.resources.getString(R.string.hardware_storage_output_format,
                            Formatter.formatFileSize(context, mi.availMem),
                            Formatter.formatFileSize(context, mi.totalMem))
                chartItem = ChartItem(mi.availMem.toFloat(), mi.totalMem.toFloat(), R.drawable.ic_memory)
            }
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun formattedInternalMemory() = Item("Internal Storage", ItemType.HARDWARE).apply {
        try {
            val dir = Environment.getDataDirectory()
            val stat = StatFs(dir.path)
            @Suppress("DEPRECATION")
            val available = if(Build.VERSION.SDK_INT <= Build.VERSION_CODES.JELLY_BEAN_MR2)
                stat.blockSize.toLong() * stat.availableBlocks.toLong()
            else stat.availableBytes
            @Suppress("DEPRECATION")
            val total = if(Build.VERSION.SDK_INT <= Build.VERSION_CODES.JELLY_BEAN_MR2)
                stat.blockCount.toLong() * stat.blockCount.toLong()
            else stat.totalBytes
            if (total > 0L)
                subtitle = context.resources.getString(R.string.hardware_storage_output_format,
                        Formatter.formatFileSize(context, available), Formatter.formatFileSize(context, total))
            chartItem = ChartItem(available.toFloat(), total.toFloat(), R.drawable.ic_storage)
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun formattedExternalMemory() = Item("External Storage", ItemType.HARDWARE).apply {
        try {
            // Mounted and not emulated, most likely a real SD Card
            val appsDir = ContextCompat.getExternalFilesDirs(context, null).filter { it != null
                    && EnvironmentCompat.getStorageState(it) == Environment.MEDIA_MOUNTED
                    && !it.isExternalStorageEmulatedCompat()}
            val availSize = appsDir.map { it.freeSpace }.toTypedArray().sum()
            val totalSize = appsDir.map { it.totalSpace }.toTypedArray().sum()
            if (totalSize > 0L)
                subtitle = context.resources.getString(R.string.hardware_storage_output_format,
                        Formatter.formatFileSize(context, availSize), Formatter.formatFileSize(context, totalSize))
            chartItem = ChartItem(availSize.toFloat(), totalSize.toFloat(), R.drawable.ic_storage)
        } catch (e: Exception) {
            Timber.w("Null in ${object{}.javaClass.enclosingMethod?.name}")
        }
    }

    private fun File.isExternalStorageEmulatedCompat(): Boolean =
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP)
            Environment.isExternalStorageEmulated(this)
        else Environment.isExternalStorageEmulated()

    private suspend fun getBattery(): Item =
            suspendCancellableCoroutine { cont ->
                val batteryReceiver = object: BroadcastReceiver() {
                    override fun onReceive(c: Context?, intent: Intent?) {
                        context.unregisterReceiver(this)
                        intent?.action?.let { act ->
                            val result = goAsync()
                            val item = Item("Battery", ItemType.HARDWARE)
                            if (act == Intent.ACTION_BATTERY_CHANGED) {
                                //int temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)/10;
                                val health = when (intent.getIntExtra(BatteryManager.EXTRA_HEALTH, 0)) {
                                    BatteryManager.BATTERY_HEALTH_COLD -> context.getString(R.string.BATTERY_HEALTH_COLD)
                                    BatteryManager.BATTERY_HEALTH_DEAD -> context.getString(R.string.BATTERY_HEALTH_DEAD)
                                    BatteryManager.BATTERY_HEALTH_GOOD -> context.getString(R.string.BATTERY_HEALTH_GOOD)
                                    BatteryManager.BATTERY_HEALTH_OVERHEAT -> context.getString(R.string.BATTERY_HEALTH_OVERHEAT)
                                    BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> context.getString(R.string.BATTERY_HEALTH_OVER_VOLTAGE)
                                    BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE -> context.getString(R.string.BATTERY_HEALTH_UNSPECIFIED_FAILURE)
                                    else -> context.getString(R.string.BATTERY_HEALTH_UNKNOWN)
                                }
                                val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                                item.subtitle = StringBuilder().apply {
                                    append("$level% - ")
                                    // Are we charging / charged?
                                    when (intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
                                        BatteryManager.BATTERY_STATUS_CHARGING -> append(context.getString(R.string.BATTERY_STATUS_CHARGING))
                                        BatteryManager.BATTERY_STATUS_FULL -> append(context.getString(R.string.BATTERY_STATUS_FULL))
                                        BatteryManager.BATTERY_STATUS_DISCHARGING -> append(context.getString(R.string.BATTERY_STATUS_DISCHARGING))
                                        BatteryManager.BATTERY_STATUS_NOT_CHARGING -> append(context.getString(R.string.BATTERY_STATUS_NOT_CHARGING))
                                        else -> append(context.getString(R.string.BATTERY_STATUS_UNKNOWN))
                                    }
                                    append(" ($health)")
                                    // How are we charging?
                                    val chargePlug = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1)
                                    val usbCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_USB
                                    val acCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_AC
                                    append(" ")
                                    if (usbCharge)
                                        append(context.getString(R.string.BATTERY_PLUGGED_USB))
                                    else if (acCharge)
                                        append(context.getString(R.string.BATTERY_PLUGGED_AC))
                                    //subTitle = String.valueOf(temperature+"\u00b0C");
                                }.toString()
                                item.chartItem = ChartItem((100 - level).toFloat(), 100f, R.drawable.ic_battery)
                                // Must call finish() so the BroadcastReceiver can be recycled.
                                result.finish()
                            }
                            cont.resume(item)
                        }
                    }
                }
                cont.invokeOnCancellation {
                    context.unregisterReceiver(batteryReceiver)
                }
                context.registerReceiver(batteryReceiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
            }

    private inner class InfoReceiver : BroadcastReceiver() {
        override fun onReceive(c: Context, intent: Intent?) {
            while(scope.isActive) {
                intent?.action?.let { act ->
                    val result = goAsync()
                    scope.launch {
                        if (act == Intent.ACTION_BATTERY_CHANGED) {
                            //int temperature = intent.getIntExtra(BatteryManager.EXTRA_TEMPERATURE, 0)/10;
                            val health = when (intent.getIntExtra(BatteryManager.EXTRA_HEALTH, 0)) {
                                BatteryManager.BATTERY_HEALTH_COLD -> context.getString(R.string.BATTERY_HEALTH_COLD)
                                BatteryManager.BATTERY_HEALTH_DEAD -> context.getString(R.string.BATTERY_HEALTH_DEAD)
                                BatteryManager.BATTERY_HEALTH_GOOD -> context.getString(R.string.BATTERY_HEALTH_GOOD)
                                BatteryManager.BATTERY_HEALTH_OVERHEAT -> context.getString(R.string.BATTERY_HEALTH_OVERHEAT)
                                BatteryManager.BATTERY_HEALTH_OVER_VOLTAGE -> context.getString(R.string.BATTERY_HEALTH_OVER_VOLTAGE)
                                BatteryManager.BATTERY_HEALTH_UNSPECIFIED_FAILURE -> context.getString(R.string.BATTERY_HEALTH_UNSPECIFIED_FAILURE)
                                else -> context.getString(R.string.BATTERY_HEALTH_UNKNOWN)
                            }
                            val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
                            val item = Item("Battery", ItemType.HARDWARE)
                            item.subtitle = StringBuilder().apply {
                                append("$level% - ")
                                // Are we charging / charged?
                                when (intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
                                    BatteryManager.BATTERY_STATUS_CHARGING -> append(context.getString(R.string.BATTERY_STATUS_CHARGING))
                                    BatteryManager.BATTERY_STATUS_FULL -> append(context.getString(R.string.BATTERY_STATUS_FULL))
                                    BatteryManager.BATTERY_STATUS_DISCHARGING -> append(context.getString(R.string.BATTERY_STATUS_DISCHARGING))
                                    BatteryManager.BATTERY_STATUS_NOT_CHARGING -> append(context.getString(R.string.BATTERY_STATUS_NOT_CHARGING))
                                    else -> append(context.getString(R.string.BATTERY_STATUS_UNKNOWN))
                                }
                                append(" ($health)")
                                // How are we charging?
                                val chargePlug = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, -1)
                                val usbCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_USB
                                val acCharge = chargePlug == BatteryManager.BATTERY_PLUGGED_AC
                                append(" ")
                                if (usbCharge)
                                    append(context.getString(R.string.BATTERY_PLUGGED_USB))
                                else if (acCharge)
                                    append(context.getString(R.string.BATTERY_PLUGGED_AC))
                                //append(" $temperature\u00b0C")
                            }.toString()
                            item.chartItem = ChartItem((100 - level).toFloat(), 100f, R.drawable.ic_battery)
                            db.itemDao().insert(item)
                            // Must call finish() so the BroadcastReceiver can be recycled.
                            result.finish()
                        }
                    }
                    context.unregisterReceiver(this)
                }
            }
        }
    }
}